File: server/src/spotify/playlists.rs
use actix_web::client::Client;
use actix_web::{HttpRequest, HttpResponse};
use serde::{Deserialize, Serialize};

#[derive(Deserialize)]
struct SpotifyImage {
    url: String,
}
#[derive(Deserialize)]
struct SpotifyPlaylist {
    description: String,
    id: String,
    images: Vec<SpotifyImage>,
    name: String,
}
#[derive(Deserialize)]
struct SpotifyPlaylistItems {
    items: Vec<SpotifyPlaylist>,
}
#[derive(Deserialize)]
struct SpotifyPlaylists {
    playlists: SpotifyPlaylistItems,
}

#[derive(Serialize)]
struct NewPlaylist {
    description: String,
    id: String,
    name: String,
    image: Option<String>,
}

pub async fn get_spotify_playlists(req: HttpRequest) -> HttpResponse {
    let bearer_token: Option<&str> = req
        .headers()
        .get("Authorization")
        .and_then(|t| Some(t.to_str().unwrap()));

    let mut playlists: Vec<NewPlaylist> = vec![];

    match bearer_token {
        Some(token) => {
            let request = Client::new()
                .get("https://api.spotify.com/v1/browse/featured-playlists?country=FR")
                .header("Authorization", token)
                .header("Accept", "application/json")
                .header("Content-Type", "application/json")
                .send()
                .await;

            match request {
                Ok(mut body) => {
                    let success: bool = body.status().as_u16() < 400;
                    if success {
                        let resp_body: SpotifyPlaylists =
                            body.json::<SpotifyPlaylists>().await.unwrap();
                        for playlist in resp_body.playlists.items {
                            playlists.push(NewPlaylist {
                                description: playlist.description,
                                id: playlist.id,
                                name: playlist.name,
                                image: playlist
                                    .images
                                    .get(0)
                                    .and_then(|i: &SpotifyImage| Some(i.url.clone())),
                            });
                        }
                    } else {
                        // Error
                        println!("response: {:?}", body);
                    }
                }
                Err(err) => {
                    println!("Error on Spotify response: {:?}", err);
                }
            };
        }
        None => {}
    };

    HttpResponse::Ok().json(playlists)
}

File: backend/src/providers/spotify/playlists.ts
import { Request, Response } from "express";
import request from "request";
import { Playlist } from "../playlist";

interface SpotifyPlaylistsResponse {
  href: string;
  items: Array<{
    description: string;
    id: string;
    images: Array<{
      url: string;
    }>;
    name: string;
  }>;
}
interface SpotifyPlaylistsBody {
  message: string;
  playlists: SpotifyPlaylistsResponse;
}

/**
 * Returns spotify playlists using user token access.
 * If a userID is given, will fetch user playlists. Spotify featured playlists Otherwise.
 *
 * @param token : string, user access token that should be: 'Bearer xxxxxxxx'.
 * @param userId : string | undefined, user id.
 */
export function requestSpotifyPlaylists(token: string, userId: string | undefined): Promise<Array<Playlist>> {
  const getOptions = {
    url: userId !== undefined ? `https://api.spotify.com/v1/users/${userId}/playlists` : "https://api.spotify.com/v1/browse/featured-playlists?country=FR",
    headers: {
      Authorization: token,
    },
    json: true,
  };

  return new Promise<Array<Playlist>>((resolve, reject) => {
    request.get(getOptions, (error, response, body: SpotifyPlaylistsBody | SpotifyPlaylistsResponse) => {
      if (!error && response.statusCode === 200) {
        const playlists: Array<Playlist> = [];
        let spotifyPlaylists: SpotifyPlaylistsResponse;
        if (userId === undefined) {
          spotifyPlaylists = (body as SpotifyPlaylistsBody).playlists;
        } else {
          spotifyPlaylists = body as SpotifyPlaylistsResponse;
        }
        for (const playlist of spotifyPlaylists.items) {
          playlists.push({
            description: playlist.description,
            id: playlist.id,
            name: playlist.name,
            image: playlist.images.length > 0 ? playlist.images[0].url : null,
          });
        }
        resolve(playlists);
      } else {
        reject();
      }
    });
  });
}

/**
 * Spotify playlists handler to get playlist at /spotify/playlists
 *
 * @param req
 * @param res
 */
export async function getSpotifyPlaylists(req: Request, res: Response): Promise<void> {
  const token: string | undefined = req.header("Authorization");
  const userId: string | undefined = req.query.userId;
  if (token === undefined) {
    throw "Bearer authorization missing !";
  }

  let playlists: Array<Playlist>;

  try {
    playlists = await requestSpotifyPlaylists(token, userId);
  } catch (e) {
    throw "Error...";
  }

  res.sendJSON(playlists);
}
